src/output-json-anomaly.c

```diff
@@ -134,6 +139,129 @@ static int AnomalyJson(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *
+                p->flow ? json_integer(p->flow->alproto) : json_string("unknown"));
+        }
+
+        const char *event_name = NULL;
+        uint8_t event_code = decoder_events->events[i];
+        AppLayerEventType event_type;
+        int r;
+        if (is_applayer) {
+            r = AppLayerGetEventInfoById(event_code, &event_name, &event_type);
+        } else {
+            r = AppLayerParserGetEventInfoById(p->flow->proto, p->flow->alproto,
+                                               event_code, &event_name, &event_type);
+        }
+        if (r == 0) {
+            json_object_set_new(ajs, "type",
+                        json_string(event_type == APP_LAYER_EVENT_TYPE_TRANSACTION ?
+                                    "transaction" : "packet"));
+            json_object_set_new(ajs, "event", json_string(event_name));
+        } else {
+            json_object_set_new(ajs, "type", json_string("unknown"));
+            json_object_set_new(ajs, "code", json_integer(event_code));
+        }
+
+        /* log event number (and total */
+        uint32_t offset = 0;
+        char event_no_buf[32];
+        PrintBufferData(event_no_buf, &offset, 32, "%d (of %d)",
+                        i+1, decoder_events->cnt);
+        json_object_set_new(ajs, "event_no", json_string(event_no_buf));
+
+        if (tx_id != (uint64_t) -1)
+            json_object_set_new(ajs, "tx_id", json_integer(tx_id));
+        json_object_set_new(ajs, "layer", json_string(layer));
+
+        /* anomaly */
+        json_object_set_new(js, ANOMALY_EVENT_TYPE, ajs);
+        OutputJSONBuffer(js, aft->file_ctx, &aft->json_buffer);
+
+        json_object_clear(js);
+        json_decref(js);
+    }
+
+    return TM_ECODE_OK;
+}
+
+static int JsonAnomalyTxLogger(ThreadVars *tv, void *thread_data, const Packet *p,
+                               Flow *f, void *state, void *tx, uint64_t tx_id)
+{
+    SCLogDebug("state %p, tx: %p, tx_id: %"PRIu64, state, tx, tx_id);
+
+    JsonAnomalyLogThread *aft = thread_data;
+    uint8_t proto = f->proto;
+    AppProto alproto = f->alproto;
+    AppLayerDecoderEvents *decoder_events;
+    decoder_events = AppLayerParserGetEventsByTx(proto, alproto, state, tx_id);
+    if (decoder_events)
+        AnomalyAppLayerDecoderEventJson(aft, p, decoder_events, false,
+                                        "applayer_parser", tx_id);
+    return TM_ECODE_OK;
+}
+
+static inline bool AnomalyHasParserEvents(const Packet *p)
+{
+    return (p->flow && p->flow->alparser &&
+            AppLayerParserHasDecoderEvents(p->flow->alparser));
+}
+
+static int AnomalyJson(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *p)
+{
+
+    int rc = TM_ECODE_OK;
+
+    if (p->events.cnt) {
+        rc = AnomalyDecodeEventJson(tv, aft, p);
+    }
+
+    /* app layer events */
+    if (p->app_layer_events && p->app_layer_events->cnt) {
+        rc = AnomalyAppLayerDecoderEventJson(aft, p, p->app_layer_events,
```

Comment:
rc is overwritten. AnomalyDecodeEventJson could have set it.

Reply:
Oops ... good catch.